python/python/bystro/covariance/_base_covariance.py

```diff
@@ -0,0 +1,549 @@
+        The predictive covariates
+
+    covariance_bar : array-like
+        Conditional covariance
+    """
+
+    covariance_sub = covariance[idxs == 1]
+    covariance_22 = covariance_sub[:, idxs == 1]
+    covariance_21 = covariance_sub[:, idxs == 0]
+    covariance_nonsub = covariance[idxs == 0]
+    covariance_11 = covariance_nonsub[:, idxs == 0]
+    beta_bar = la.solve(covariance_22, covariance_21)
+    Second_part = np.dot(covariance_21.T, beta_bar)
+
+    covariance_bar = covariance_11 - Second_part
+    return beta_bar.T, covariance_bar
+
+
+def _marginal_score(covariance, X, idxs, weights=None):
+    """
+    Returns the marginal log-likelihood of a subset of data
+
+    Parameters
+    ----------
+    covariance : np.array-like(p,p)
+        The covariance matrix
+
+    X : np.array-like,(N,sum(idxs))
+        The data
+
+    idxs: np.array-like,(sum(p),)
+        The observation locations
+
+    weights : np.array-like,(N,),default=None
+        The optional weights on the samples
+
+    Returns
+    -------
+    avg_score : float
+        Average log likelihood
+    """
+    if weights is None:
+        weights = np.ones(X.shape[0])
+    avg_score = np.mean(weights * _marginal_score_samples(covariance, X, idxs))
+    return avg_score
+
+
+def _marginal_score_samples(covariance, X, idxs):
+    """
+    Returns the marginal log-likelihood of a subset of data
+
+    Parameters
+    ----------
+    covariance : np.array-like(p,p)
+        The covariance matrix
+
+    X : np.array-like,(N,sum(idxs))
+        The data
+
+    idxs: np.array-like,(sum(p),)
+        The observation locations
+
+    Returns
+    -------
+    scores : float
+        Average log likelihood
+    """
+    cov1 = covariance[idxs == 1]
+    cov_sub = cov1[:, idxs == 1]
+    scores = _score_samples(cov_sub, X)
+    return scores
+
+
+def _score(covariance, X, weights=None):
+    """
+    Returns the average log liklihood of data.
+
+    Parameters
+    ----------
+    covariance : np.array-like(p,p)
+        The covariance matrix
+
+    X : np.array-like,(N,sum(p))
+        The data
+
+    weights : np.array-like,(N,),default=None
+        The optional weights on the samples
+
+    Returns
+    -------
+    avg_score : float
+        Average log likelihood
+    """
+    if weights is None:
+        weights = np.ones(X.shape[0])
+    avg_score = np.mean(weights * _score_samples(covariance, X))
+    return avg_score
+
+
+def _score_samples(covariance, X):
+    """
+    Return the log likelihood of each sample
+
+    Parameters
+    ----------
+    covariance : np.array-like(p,p)
+        The covariance matrix
+
+    X : np.array-like,(N,sum(p))
+        The data
+
+    Returns
+    -------
+    scores : float
+        Log likelihood for each sample
+    """
+    p = covariance.shape[0]
+    term1 = -p / 2 * np.log(2 * np.pi)
+
+    _, logdet = la.slogdet(covariance)
+    term2 = -0.5 * logdet
+
+    quad_init = la.solve(covariance, np.transpose(X))
```

Comment:
could we just get a brief gloss of what's going on in these three lines?  AFAICT we're just calculating the log-density of X, but I think the modal engineer would have to stop and think for a while about why this works.

Also, here and throughout, are we guaranteed that X is already centered?

Reply:
Yes

Reply:
Or more accurately that's the requirement, maybe should state explicitly.

Reply:
I think it would really help to state that explicitly somewhere, because we're making a lot of assumptions that X is centered throughout the code but AFAICT never checking it explicitly.